pycontrails/datalib/ecmwf/model_level_hres.py

```diff
@@ -0,0 +1,511 @@
+from pycontrails.datalib.ecmwf.common import ECMWFAPI
+from pycontrails.datalib.ecmwf.model_levels import pressure_levels_at_model_levels
+from pycontrails.utils import dependencies, temp
+from pycontrails.utils.types import DatetimeLike
+
+MODEL_LEVEL_VARIABLES = [
+    met_var.AirTemperature,
+    met_var.SpecificHumidity,
+    met_var.VerticalVelocity,
+    met_var.EastwardWind,
+    met_var.NorthwardWind,
+    ecmwf_var.RelativeVorticity,
+    ecmwf_var.Divergence,
+    ecmwf_var.CloudAreaFractionInLayer,
+    ecmwf_var.SpecificCloudIceWaterContent,
+    ecmwf_var.SpecificCloudLiquidWaterContent,
+]
+
+LAST_STEP_1H = 96  # latest forecast step with 1 hour frequency
+LAST_STEP_3H = 144  # latest forecast step with 3 hour frequency
+LAST_STEP_6H = 240  # latest forecast step with 6 hour frequency
+
+
+class ModelLevelHRES(ECMWFAPI):
+    """Class to support model-level HRES data access, download, and organization.
+
+    The interface is similar to :class:`pycontrails.datalib.ecmwf.HRES`,
+    which downloads pressure-level data with much lower vertical resolution and single-level data.
+    Note, however, that only a subset of the pressure-level data available through the operational
+    archive is available as model-level data. As a consequence, this interface only
+    supports access to nominal HRES forecasts (corresponding to ``stream = "oper"`` and
+    ``field_type = "fc"`` in :class:`pycontrails.datalib.ecmwf.HRES`) initialized at 00z and 12z.
+
+    Requires account with ECMWF and API key.
+
+    API credentials can be set in local ``~/.ecmwfapirc`` file:
+
+    .. code:: json
+
+        {
+            "url": "https://api.ecmwf.int/v1",
+            "email": "<email>",
+            "key": "<key>"
+        }
+
+    Credentials can also be provided directly in ``url``, ``key``, and ``email`` keyword args.
+
+    See `ecmwf-api-client <https://github.com/ecmwf/ecmwf-api-client>`_ documentation
+    for more information.
+
+    Parameters
+    ----------
+    time : datalib.TimeInput
+        The time range for data retrieval, either a single datetime or (start, end) datetime range.
+        Input must be datetime-like or tuple of datetime-like
+        (:py:class:`datetime.datetime`, :class:`pandas.Timestamp`, :class:`numpy.datetime64`)
+        specifying the (start, end) of the date range, inclusive.
+        All times will be downloaded in a single GRIB file, which
+        ensures that exactly one request is submitted per file on tape accessed.
+        If ``forecast_time`` is unspecified, the forecast time will
+        be assumed to be the nearest synoptic hour available in the operational archive (00 or 12).
+        All subsequent times will be downloaded for relative to :attr:`forecast_time`.
+    variables : datalib.VariableInput
+        Variable name (i.e. "t", "air_temperature", ["air_temperature, specific_humidity"])
+    pressure_levels : datalib.PressureLevelInput, optional
```

Comment:
Is it worth keeping the `pressure levels` input to these classes?  Would it work if I passed in arbitrary pressure levels that were a different length than the `levels` input?

Reply:
Yes! `pressure_levels` can be anything you want, with the caveat that you might get some NaNs if you set the `levels` input to something that doesn't download enough model levels to cover the range of pressure levels you request.